Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -736,8 +733,23 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
                     auto const normal_dot_vertex_to_light = vertex.normal.dot(FloatVector3(vertex_to_light.x(), vertex_to_light.y(), vertex_to_light.z()));
                     auto const diffuse_component = ((diffuse * light.diffuse_intensity) * normal_dot_vertex_to_light).clamped(0.0f, 1.0f);
 
+                    // Specular
+                    FloatVector4 specular_component = { 0.0f, 0.0f, 0.0f, 0.0f };
+                    if (normal_dot_vertex_to_light > 0.0f) {
+                        FloatVector4 half_vector_normalized;
+                        if (!m_lighting_model.viewer_at_infinity)
+                            half_vector_normalized = (vertex_to_light + FloatVector4(0.0f, 0.0f, 1.0f, 0.0f)).normalized();
+                        else
+                            half_vector_normalized = (vertex_to_light + FloatVector4(0.0f, 0.0f, 0.0f, 1.0f)).normalized();
```

Comment:
According to GL1.5 spec, the right hand vector is `VPe` i.e the vector from the vertex's position to the eye point. If we use the lambda expression for that as commented earlier this will become much clearer.